File: examples/python/mongodb/mongodb_vfs.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import json
import os

from dotenv import load_dotenv

from mirage import MountMode, Workspace
from mirage.resource.mongodb import MongoDBConfig, MongoDBResource

load_dotenv(".env.development")

DB = "mirage_test"
COLL = "heterogeneous"
VIEW = "high_rated_films"

config = MongoDBConfig(uri=os.environ["MONGODB_URI"], databases=[DB])
resource = MongoDBResource(config=config)


async def main():
    with Workspace({"/mongodb/": resource}, mode=MountMode.READ) as ws:
        print("=== VFS MODE: open() reads from MongoDB ===\n")

        print("--- listdir() root (databases) ---")
        for db in os.listdir("/mongodb"):
            print(f"  {db}")

        print(f"\n--- listdir() /mongodb/{DB} (entities) ---")
        for entry in os.listdir(f"/mongodb/{DB}"):
            print(f"  {entry}")

        print(f"\n--- listdir() /mongodb/{DB}/collections ---")
        collections = os.listdir(f"/mongodb/{DB}/collections")
        for col in collections:
            print(f"  {col}")

        print(f"\n--- listdir() /mongodb/{DB}/views ---")
        views = os.listdir(f"/mongodb/{DB}/views")
        for v in views:
            print(f"  {v}")

        print(f"\n--- listdir() /mongodb/{DB}/collections/{COLL} (entity) ---")
        for entry in os.listdir(f"/mongodb/{DB}/collections/{COLL}"):
            print(f"  {entry}")

        print("\n--- open() database.json ---")
        with open(f"/mongodb/{DB}/database.json") as f:
            db_meta = json.loads(f.read())
        print(f"  collections: {len(db_meta.get('collections', []))}")
        print(f"  views: {len(db_meta.get('views', []))}")

        print(f"\n--- open() schema.json for {COLL} ---")
        with open(f"/mongodb/{DB}/collections/{COLL}/schema.json") as f:
            schema = json.loads(f.read())
        print(f"  kind: {schema.get('kind')}")
        print(f"  fields: {len(schema.get('fields', []))}")
        print(f"  indexes: {len(schema.get('indexes', []))}")

        print(f"\n--- open() + read documents.jsonl for {COLL} ---")
        path = f"/mongodb/{DB}/collections/{COLL}/documents.jsonl"
        with open(path) as f:
            content = f.read()
        lines = [ln for ln in content.strip().split("\n") if ln.strip()]
        print(f"  documents: {len(lines)}")
        for line in lines[:3]:
            doc = json.loads(line)
            print(f"  [{doc['_id']['$oid']}] {doc.get('title', '?')}")

        print(f"\n--- open() + read view documents for {VIEW} ---")
        view_path = f"/mongodb/{DB}/views/{VIEW}/documents.jsonl"
        with open(view_path) as f:
            view_content = f.read()
        view_lines = [
            ln for ln in view_content.strip().split("\n") if ln.strip()
        ]
        print(f"  documents: {len(view_lines)}")
        for line in view_lines[:3]:
            doc = json.loads(line)
            title = doc.get("title", "?")
            rating = doc.get("rating", "?")
            print(f"  {title} (rating={rating})")

        print("\n--- bash history ---")
        with open("/.bash_history") as f:
            for i, line in enumerate(f):
                if i >= 6:
                    break
                print(f"  {line.rstrip()[:120]}")

        records = ws.fs.records
        total = sum(r.bytes for r in records)
        print(f"\nStats: {len(records)} ops, "
              f"{total} bytes transferred")


asyncio.run(main())

File: examples/typescript/mongodb/mongodb_vfs.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { setServers } from 'node:dns'
import dotenv from 'dotenv'
import { MongoDBResource, MountMode, Workspace } from '@struktoai/mirage-node'

setServers(['8.8.8.8', '1.1.1.1'])
dotenv.config({ path: '.env.development' })

const uri = process.env.MONGODB_URI
if (uri === undefined) {
  console.error('MONGODB_URI missing in .env.development')
  process.exit(1)
}

const DB = 'mirage_test'
const COLL = 'heterogeneous'
const VIEW = 'high_rated_films'

const DEC = new TextDecoder()

async function dump(ws: Workspace, label: string, cmd: string): Promise<void> {
  console.log(`\n--- ${label} ---`)
  const r = await ws.execute(cmd)
  if (r.exitCode !== 0) {
    console.log(`(exit=${String(r.exitCode)}) ${DEC.decode(r.stderr)}`)
    return
  }
  const out = DEC.decode(r.stdout)
  for (const ln of out.trimEnd().split('\n').slice(0, 8)) console.log(`  ${ln.slice(0, 160)}`)
}

async function main(): Promise<void> {
  const resource = new MongoDBResource({
    uri,
    databases: [DB],
  })
  const ws = new Workspace({ '/mongodb/': resource }, { mode: MountMode.READ })

  try {
    console.log('=== VFS MODE: shell pipelines transparently read MongoDB ===')

    const base = `/mongodb/${DB}`
    const collDoc = `${base}/collections/${COLL}/documents.jsonl`
    const collSchema = `${base}/collections/${COLL}/schema.json`
    const viewDoc = `${base}/views/${VIEW}/documents.jsonl`

    await dump(ws, 'listdir root', 'ls /mongodb/')
    await dump(ws, `listdir ${base}/`, `ls ${base}/`)
    await dump(ws, `listdir ${base}/collections/`, `ls ${base}/collections/`)
    await dump(ws, `listdir ${base}/views/`, `ls ${base}/views/`)
    await dump(
      ws,
      `listdir ${base}/collections/${COLL}/ (entity)`,
      `ls ${base}/collections/${COLL}/`,
    )

    await dump(ws, `cat ${base}/database.json`, `cat ${base}/database.json`)
    await dump(ws, `cat schema.json for ${COLL}`, `cat ${collSchema}`)
    await dump(ws, `head -n 3 documents.jsonl`, `head -n 3 ${collDoc}`)
    await dump(ws, `wc -l documents.jsonl`, `wc -l ${collDoc}`)
    await dump(ws, `head -n 2 view documents`, `head -n 2 ${viewDoc}`)
    await dump(ws, `jq titles`, `jq -r ".[] | .title" ${collDoc} | head -n 5`)
  } finally {
    await ws.close()
    await resource.close()
  }
}

await main()
